### ISSUE TITLE
alternator_stream_tests.py::TestAlternatorStreams::test_verify_all_nodes_have_same_stream fails

### ISSUE THREAD

**Issue by tchaikov** [scylladb/scylladb#15039]
see https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/3719/

```
[2023-08-14T03:54:02.919Z] self = <alternator_stream_tests.TestAlternatorStreams object at 0x7f77a8e31350>
[2023-08-14T03:54:02.919Z] 
[2023-08-14T03:54:02.919Z]     @pytest.mark.next_gating
[2023-08-14T03:54:02.919Z]     def test_verify_all_nodes_have_same_stream(self):
[2023-08-14T03:54:02.919Z]         num_of_items = NUM_OF_ITEMS
[2023-08-14T03:54:02.919Z]         self.prepare_dynamodb_cluster(num_of_nodes=3)
[2023-08-14T03:54:02.919Z]         node1 = self.cluster.nodelist()[0]
[2023-08-14T03:54:02.919Z] >       stream_arn = self.prefill_dynamodb_table(
[2023-08-14T03:54:02.919Z]             node=node1, stream_specification=enums.StreamSpecification.KEYS_ONLY.value, num_of_items=num_of_items)[0]
[2023-08-14T03:54:02.919Z] 
[2023-08-14T03:54:02.919Z] alternator_stream_tests.py:25: 
[2023-08-14T03:54:02.919Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2023-08-14T03:54:02.919Z] alternator_utils.py:876: in prefill_dynamodb_table
[2023-08-14T03:54:02.919Z]     table = super().prefill_dynamodb_table(node=node, table_name=table_name, num_of_items=num_of_items, **kwargs)
[2023-08-14T03:54:02.919Z] alternator_utils.py:723: in prefill_dynamodb_table
[2023-08-14T03:54:02.919Z]     return self.batch_write_actions(table_name=table_name, node=node, new_items=new_items)
[2023-08-14T03:54:02.919Z] alternator_utils.py:395: in batch_write_actions
[2023-08-14T03:54:02.919Z]     raise error
[2023-08-14T03:54:02.919Z] alternator_utils.py:388: in batch_write_actions
[2023-08-14T03:54:02.919Z]     batch.put_item(item)
[2023-08-14T03:54:02.919Z] /usr/local/lib/python3.11/site-packages/boto3/dynamodb/table.py:103: in put_item
[2023-08-14T03:54:02.919Z]     self._add_request_and_process({'PutRequest': {'Item': Item}})
[2023-08-14T03:54:02.919Z] /usr/local/lib/python3.11/site-packages/boto3/dynamodb/table.py:112: in _add_request_and_process
[2023-08-14T03:54:02.919Z]     self._flush_if_needed()
[2023-08-14T03:54:02.919Z] /usr/local/lib/python3.11/site-packages/boto3/dynamodb/table.py:139: in _flush_if_needed
[2023-08-14T03:54:02.919Z]     self._flush()
[2023-08-14T03:54:02.919Z] /usr/local/lib/python3.11/site-packages/boto3/dynamodb/table.py:144: in _flush
[2023-08-14T03:54:02.919Z]     response = self._client.batch_write_item(
[2023-08-14T03:54:02.919Z] /usr/local/lib/python3.11/site-packages/botocore/client.py:535: in _api_call
[2023-08-14T03:54:02.919Z]     return self._make_api_call(operation_name, kwargs)
[2023-08-14T03:54:02.919Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
...
[2023-08-14T03:54:02.920Z]         if http.status_code >= 300:
[2023-08-14T03:54:02.920Z]             error_code = parsed_response.get("Error", {}).get("Code")
[2023-08-14T03:54:02.920Z]             error_class = self.exceptions.from_code(error_code)
[2023-08-14T03:54:02.920Z] >           raise error_class(parsed_response, operation_name)
[2023-08-14T03:54:02.921Z] E           botocore.errorfactory.InternalServerError: An error occurred (InternalServerError) when calling the BatchWriteItem operation (reached max retries: 0): Internal server error: exceptions::invalid_request_exception (cdc: attempted to get a stream from an earlier generation than the currently used one. With CDC you cannot send writes with timestamps too far into the past, because that would break consistency properties (write timestamp: 2023/08/14 03:39:36, current generation started at: 2023/08/14 03:39:36))
```

**Comment by avikivity** [comment]
Seen again:

scylla-master/next failed on Sanity Tests stage, for more details: https://jenkins.scylladb.com/job/scylla-master/job/next/6392/
failed tests: 3 of 22612.
- alternator_stream_tests.TestAlternatorStreams.test_verify_all_nodes_have_same_stream
    botocore.errorfactory.InternalServerError: An error occurred (InternalServerError) when calling the BatchWriteItem operation (reached max retries: 0): Internal server error: exceptions::invalid_request_exception (cdc: attempted to get a stream from an earlier generation than the currently used one. With CDC you cannot send writes with timestamps too far into the past, because that would break consistency properties (write timestamp: 2023/08/14 07:31:24, current generation started at: 2023/08/14 07:31:24))
- test_tablets.test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode
    cassandra.OperationTimedOut: errors={'127.115.92.3:9042': 'Client request timeout. See Session.execute[_async](timeout)'}, last_host=127.115.92.3:9042
- non-boost tests.topology_experimental_raft.test_tablets.debug.1
Full details on the Test Result page: https://jenkins.scylladb.com/job/scylla-master/job/next/6392/testReport/

**Comment by nyh** [comment]
The relevant error message is:

> botocore.errorfactory.InternalServerError: An error occurred (InternalServerError) when calling the BatchWriteItem operation (reached max retries: 0): Internal server error: exceptions::invalid_request_exception (**cdc: attempted to get a stream from an earlier generation than the currently used one. With CDC you cannot send writes with timestamps too far into the past, because that would break consistency properties (write timestamp: 2023/08/14 03:39:36, current generation started at: 2023/08/14 03:39:36)**)

The two printed timestamps appear the same, certainly not "too far into the past".

**Comment by nyh** [comment]
I suspect the problem is the recently reduced sleeps when starting nodes: Whereas in the past we had a long sleep after starting a node before we start writing, today tests start writing faster when nodes appear to be all available and aware of each other.

@kbr-scylla @elcallio  is it possible that when a Scylla cluster boots, there is some period (maybe 10 seconds? some other constant?) when writes with CDC are not allowed because the generation is "too new"? Are clients supposed to wait a bit before enabling CDC on a new cluster? How do they know how long to wait?

**Comment by nyh** [comment]
I think this should be considered is a dtest bug:

In https://github.com/scylladb/scylladb/issues/6864 we noticed that Alternator writes with CDC enabled indeed fail if done too early. In https://github.com/scylladb/scylladb/pull/6901, already three years ago, @elcallio fixed it so the stream will be marked unavailable until that time. This means that if an application enables stream, and then **waits** for it to be enabled (as we assumed users and tests would do), everything would be fine. This isn't pretty, but we decided that it's acceptable and not considered a bug.

It seems that the dtest Alternator convenience functions, tried to do this, but tried to do it wrong: In alternator_utils.py there is
```python
    def prefill_dynamodb_table(self, node: ScyllaNode, table_name: str = TABLE_NAME, num_of_items: int = NUM_OF_ITEMS,
                               wait_for_active_stream=True, **kwargs):
        stream_arn_details = None
        table = super().prefill_dynamodb_table(node=node, table_name=table_name, num_of_items=num_of_items, **kwargs)
        if wait_for_active_stream:
            stream_arn_details = self.wait_for_active_stream(node=node, table_name=table.name)
        return stream_arn_details
```

The problem is that this function first writes to the table, and only waits for the enabled stream to become active. This is a mistake, the order should have been reversed! It should first wait for the stream to become active, and only then try writing. If we do it in the current (reversed) way we risk some of the written data not being saved to the stream, and - as we saw in this issue - even some of the writes failing.



But this test doesn't do it: As far as I can tell enables the stream right when creating the table, and then does not wait for the stream to be fully enabled - and start writing data immediately.

I think this is a test bug for two reasons:
1. It can cause the writes to start before the generation is available (the problem we saw in this failure)
2. Even if not, theoretically it may cause some of the writes to not be saved in the CDC (because it's not fully enabled yet). This specific test will not fail (it doesn't check if the result of reading from the CDC is correct!) but could break other tests.
